### PR TITLE
Patrick customizable app container

### PR DESCRIPTION
**Pull request by patrick-tavares**
This PR addresses the new functionality requested in issue #2874.

Ultimately the core <App> component building code was moved out from imports/plugins/core/router/client/app.js into a registerable component in imports/plugins/core/ui/clients/components/apps and imports/plugins/core/ui/clients/containers (though I am open to guidance if there is a better place for this to live). In addition, defaults.js is used to specify what the overriding AppContainer should be.

### CODE REVIEW

**Comment by patrick-tavares** [comment]
So one thing I just noticed in going through another round of UI tests is that, while double-clicking a product grid item (while logged in as admin) properly displays the PDP page with Product Settings ActionView in FireFox, Safari and Chrome keep showing the Product Grid page with the Grid Settings ActionView. I could use another set of eyes here as the productGridItemsContainer.js code wasn't touched but seems to not like my changes.

File: client/config/defaults.js [hunk 1]
```diff
@@ -7,6 +7,10 @@ import { AutoForm } from "meteor/aldeed:autoform";
  * A place to put misc. package configurations
  */
 
+// Set the default React app container inside which the rest of the UI will live
+// This enables the ability to completely override the look and feel of RC
+export const DEFAULT_APP_CONTAINER = "CoreReactionApp";
```
Comment (mikemurray) [inline client/config/defaults.js]: What's the need to set a session variable for the default app container?

Could you not just override it by using `replaceComponent("CoreReactionApp", MyCustomApp);`

Comment (patrick-tavares) [reply client/config/defaults.js]: I don't think I have a good response here. I'd gotten used to seeing registerComponent in digging through the code. Going the replaceComponent route simplifies things a bit so I'm onboard.

File: imports/plugins/core/ui/client/components/app/coreReactionApp.js [hunk 2]
```diff
@@ -0,0 +1,106 @@
+import { Switch } from "react-router-dom";
```
Comment (mikemurray) [inline imports/plugins/core/ui/client/components/app/coreReactionApp.js]: Simply call is app.js and class name App. core and reaction are not necessary. Everywhere its called CoreReactionApp should just be App.

File: imports/plugins/core/router/client/app.js [hunk 3]
```diff
@@ -1,116 +1,15 @@
...
+import { Session } from "meteor/session";
+import { getComponent } from "@reactioncommerce/reaction-components";
+
+const App = (props) => {
+  // Make sure a default value is set.
+  const defaultContainer = Session.get("DEFAULT_APP_CONTAINER") || "CoreReactionApp";
+  const AppContainer = getComponent(defaultContainer);
+
+  return (
+    <AppContainer {...props} />
```
Comment (mikemurray) [inline imports/plugins/core/router/client/app.js]: This file doesn't need to exist anymore. In `browserRouter.js` file in the same folder, you could just do the following.

```
import { Components } from "@reactioncommerce/reaction-components";

<Components.App>
```

File: client/config/defaults.js [hunk 4]
```diff
@@ -22,6 +26,7 @@ export const INDEX_OPTIONS = {
 // default load qty for product grid
 export const ITEMS_INCREMENT = 24;
 
+Session.setDefault("DEFAULT_APP_CONTAINER", DEFAULT_APP_CONTAINER);
```
Comment (mikemurray) [inline client/config/defaults.js]: Don't use session vars in `default.js` we will most likely move away from this solution in the future. And you can override components using `replaceComponent`

File: imports/plugins/core/ui/client/components/index.js [hunk 5]
```diff
@@ -1,6 +1,7 @@
 // export ButtonGroup from "./buttonGroup/buttonGroup";
 
 export { Alerts, Alert } from "./alerts";
+export { CoreReactionApp } from "./app";
```
Comment (mikemurray) [inline imports/plugins/core/ui/client/components/index.js]: export as `App`

File: imports/plugins/core/ui/client/containers/index.js [hunk 6]
```diff
@@ -1,5 +1,6 @@
 export { default as EditContainer } from "./edit";
 export { default as Alerts } from "./alerts";
+export { default as CoreReactionAppContainer } from "./coreReactionAppContainer";
```
Comment (mikemurray) [inline imports/plugins/core/ui/client/containers/index.js]: We can just call it App, no need for the container in the name.

**Pull request merged**